Make services non-static

bvisness opened the pull request:
This one is huge. All services have been refactored to not use static, and all dependencies for those services are now handled through the injector. All GET domains now require their specific service in their
constructor, which is what allows the injector to work its magic in the first place.

Each endpoint has been tested...somewhat thoroughly. More thorough testing will happen when other refactoring tasks are complete.

I really recommend just looking at the code from scratch for this one. The diff is insane. Sorry it's not easier to follow, but refactoring this touches most of the files in the project.

Hopefully resolves #2!

bvisness committed:
Make services non-static

This one is huge. All services have been refactored to not use static,
and all dependencies for those services are now handled through the
injector. All GET domains now require their specific service in their
constructor, which is what allows the injector to work its magic in
the first place.

Each endpoint has been tested...somewhat thoroughly. More thorough
testing will happen when other refactoring tasks are complete.

shadowhand commented on src/Domain/FeedbackGetDomain.php:
```diff
@@ -2,13 +2,19 @@
 
 abstract class FeedbackGetDomain extends FeedbackDomain
 {
     protected $redis;
```
I think this one can be private too, if the only access to redis is via `getRedisValue` and `setRedisValue`.

bvisness replied:
Good catch

shadowhand commented on src/Service/BlogService.php:
```diff
@@ -5,10 +5,12 @@
-        $wp = new WordpressClient($_ENV['WP_ENDPOINT'], $_ENV['WP_USER'], $_ENV['WP_PASSWORD']);
-        return $wp;
+        $url = rtrim($blogUrl, "/") . "/xmlrpc.php";
+        $this->client = new WordpressClient($url, $username, $password);
```
This isn't really what I meant. The `BlogService` should consume a fully configured `WordpressClient` instead of creating one. Any time you have `new Foo`, it means you're bypassing dependency injection and breaking dependency inversion.

bvisness replied:
Ok, sorry I misunderstood that. What about services that don't rely on another class and make the calls themselves? Should they stay as they are?

shadowhand replied:
Yes. The goal is to never have `new Foo` anywhere outside of the injector configuration, because it can't be tested.

shadowhand commented on src/Service/FacebookService.php:
```diff
@@ -5,44 +5,42 @@
+        $this->fb = new Facebook([
+            'app_id' => $app_id,
+            'app_secret' => $app_secret,
+            'default_graph_version' => 'v2.4'
```
Same situation as Wordpress here.

shadowhand commented on web/index.php:
```diff
@@ -2,14 +2,71 @@
+    );
+});
+$injector->delegate("$services\AppStoreService", function(EnvLoader $loader) {
+    $env = $loader->toArray();
```
Having to repeat this isn't very useful. I would either do:

``` php
$env = $loader->toArray();
// function () use ($env) { ... }
```

Or:

``` php
$injector->prepare(EnvLoader::class, function (EnvLoader $loader) {
    $loader->parse->putenv();
});
// getenv('foo')
```

shadowhand commented on web/index.php:
```diff
@@ -2,14 +2,71 @@
+$injector->delegate("$services\TwitterService", function(EnvLoader $loader) {
+    $env = $loader->toArray();
+    return new Wheniwork\Feedback\Service\TwitterService(
+        $_ENV['TWITTER_OAUTH_TOKEN'],
```
This is still using global `$_ENV` and probably doesn't work.

shadowhand commented:
Having unit tests would make it easier to prove all of this. :)

bvisness commented:
I addressed all of your comments; take another look and let me know what you think.

shadowhand commented on src/Domain/GetAppStore.php:
```diff
@@ -22,7 +37,7 @@ protected function getOutputKeyName()
     protected function getFeedbackItems()
     {
-        return AppStoreService::getReviews($_ENV['WIW_IOS_APP_ID'], $this->getRedisValue());
+        return $this->appStore->getReviews($_ENV['WIW_IOS_APP_ID'], $this->getRedisValue());
```
`$_ENV` being used here.

shadowhand commented on src/Domain/GetTwitter.php:
```diff
@@ -51,14 +65,14 @@ protected function getFeedbackHTML($feedbackItem)
     private function getTweetsSince($last_id) {
-        return TwitterService::get('https://api.twitter.com/1.1/statuses/user_timeline.json', [
+        return $this->twitter->get('https://api.twitter.com/1.1/statuses/user_timeline.json', [
             'screen_name' => $_ENV['TWITTER_WATCH_NAME'],
```
Also here.

shadowhand commented on src/Service/BlogService.php:
```diff
@@ -5,10 +5,11 @@
-    private static function getClient()
+    private $client;
+
+    public function __construct(WordpressClient $client)
```
:+1:

shadowhand commented on web/index.php:
```diff
@@ -2,14 +2,81 @@
+        'default_graph_version' => 'v2.4'
+    ]);
+    $service = new Wheniwork\Feedback\Service\FacebookService($fb, getenv('FB_PAGE_ID'));
+    $service->authenticate(
```
You might want to do a similar thing with Twitter, if possible.

bvisness replied:
What specifically? The authenticate method? (I can't tell from the diff)

shadowhand replied:
Yes, authenticating with twitter as part of the service delegation.

shadowhand commented on web/index.php:
```diff
@@ -2,14 +2,81 @@
+        'app_secret' => getenv('FB_APP_SECRET'),
+        'default_graph_version' => 'v2.4'
+    ]);
+    $service = new Wheniwork\Feedback\Service\FacebookService($fb, getenv('FB_PAGE_ID'));
```
Technically, the service factories should also be pulled from the injector to allow overloading. But since this is all happening in the front controller, it's okay to do it this way.

shadowhand commented:
Looking much better! A couple of stray env things to clean up.

bvisness committed:
Remove all instances of $_ENV

In some places, this required some more significant refactoring.
All .env-related stuff is now loaded through the injector now though!
(awyeah)

bvisness commented:
$_ENV has been purged from the project and all environment variables are loaded in through the injector now!

shadowhand commented on src/Domain/DoGeneric.php:
```diff
@@ -2,18 +2,30 @@
-            } else if ($input['key'] != $_ENV['POST_KEY']) {
-                throw new RuntimeException("The provided authentication key was invalid.");
+            if (! $this->auth->checkInput($input)) {
+                throw new RuntimeException("Unable to authenticate.");
```
Alternatively, you could also do something like:

``` php
try {
    $this->auth->ensure($input); // no if/else !
    ...
}
```

Assuming that:

``` php
class Authorizer
{
    // ...
    public function ensure(array $input) {
    {
        if (empty($input['key']) {
            throw new RuntimeException('You must provide a key with your request');
        }
        if ($input['key'] !== $this->key) {
            throw new RuntimeException('The provided authentication key was invalid');
        }
    }
}
```

This way you still have separate messages for each condition.

bvisness replied:
Ooooh, yes

shadowhand commented on src/Domain/DoZendesk.php:
```diff
@@ -2,18 +2,30 @@
+        Authorizer $auth
+    ) {
+        parent::__construct($hipchat, $github);
+        $this->auth = $auth;
```
Looks like you're repeating the auth assignment, so maybe put the auth in the `FeedbackDomain`? If you did so, you could keep the `private $auth` by providing a method in the parent to run the auth call:

``` php
try {
    $this->authorize($input);
    ...
}
```

shadowhand commented on web/index.php:
```diff
@@ -2,14 +2,98 @@
+// Services
+$injector->delegate("$services\Authorizer", function() {
+    return new Wheniwork\Feedback\Service\Authorizer(
+        getenv('POST_KEY')
```
Perfect!

shadowhand commented:
Totally up to you if you want to make the additional changes I just suggested. If not, merge away!

bvisness committed:
Refactor POST domains

POST-related domains have been refactored like the GET domains.

bvisness commented:
One last code review (hopefully)...I refactored the POST-related domains like the GET domains, which allows me to not repeat the auth assignment and to make things more consistent in general.

bvisness commented:
That last one may be outside the scope of this PR though...?

shadowhand commented on src/Domain/FeedbackPostDomain.php:
```diff
@@ -0,0 +1,99 @@
+
+            foreach ($this->getRequiredFields() as $field) {
+                if (empty($input[$field])) {
+                    throw new \RuntimeException("Missing required field '$field'");
```
This is okay but does have a downside in only showing a single error at a time. Often I do something like this:

``` php
$missing = array_diff($this->getRequiredFields(), array_keys($input));
if ($missing) {
    throw new RuntimeException(
        'Missing some required values: ' .
        implode(', ', $missing)
    );
}
```

This reduces the number of times a person has to call the code to see multiple missing values.

bvisness replied:
Ooh, I didn't even know about that! Thanks, I'll change that.

shadowhand commented on src/Service/TwitterService.php:
```diff
@@ -36,4 +33,23 @@ public static function get($request, $params = []) {
+    public function getTweetURL($tweet) {
+        $screen_name = $tweet->user->screen_name;
+        $id_str = $tweet->id_str;
+        return "https://twitter.com/$screen_name/status/$id_str";
```
You could also use `sprintf` to interpolate the strings. Matter of personal preference.

shadowhand commented:
Looks great!
